### PR TITLE
Better inform users of migration failure reason when running Apple Silicon build

### PR DESCRIPTION
**Pull request by peppy**
As mentioned in https://github.com/ppy/osu/discussions/17409#discussioncomment-2445464?

### CODE REVIEW

**Review by bdach — DISMISSED** [review]

File: osu.Game/Database/EFToRealmMigrator.cs [hunk 1]
```diff
@@ -147,19 +152,33 @@ private void beginMigration()
...
                 else
                 {
                     log("Migration failed!");
                     Logger.Log(t.Exception.ToString(), LoggingTarget.Database);
 
+                    if (RuntimeInfo.OS == RuntimeInfo.Platform.macOS && t.Exception.Flatten().InnerException is TypeInitializationException)
+                    {
+                        // Not guaranteed to be the only cause of exception, but let's roll with it for now.
+                        log("Please download and run the intel version of osu! once\nto allow data migration to complete!");
+                        return;
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Database/EFToRealmMigrator.cs:170]: Never mind, I think my testing may have been insufficient as I threw too early in the process. Will re-test.

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Database/EFToRealmMigrator.cs:171]: Have re-tested with

```diff
diff --git a/osu.Game/Database/OsuDbContext.cs b/osu.Game/Database/OsuDbContext.cs
index 79183b6f0e..a71334a084 100644
--- a/osu.Game/Database/OsuDbContext.cs
+++ b/osu.Game/Database/OsuDbContext.cs
@@ -42,8 +42,7 @@ static OsuDbContext()
             // required to initialise native SQLite libraries on some platforms.
             Batteries_V2.Init();
 
-            // https://github.com/aspnet/EntityFrameworkCore/issues/9994#issuecomment-508588678
-            raw.sqlite3_config(2 /*SQLITE_CONFIG_MULTITHREAD*/);
+            throw new TypeInitializationException("a", new Exception());
         }
 
         /// <summary>

```

**Review by bdach — DISMISSED** [review]
seems okay here, gonna wait for a 2nd from a macOS-owning person

**Comment by frenzibyte** [comment]
There is also another EF context retrieval in `RealmContextFactory.getRulesetShortNameFromLegacyID`, causing the game to crash immediately on startup:
CleanShot 2022-03-28 at 18 45 43@2x

Catching such exception to present the message in this PR gets a bit ugly as it involves catching in `getRealmInstance` and `Compact` as well, and returning an empty realm instance to keep things going:
```diff
diff --git a/osu.Game/Database/RealmAccess.cs b/osu.Game/Database/RealmAccess.cs
index 8574002436..d3a4064cc8 100644
--- a/osu.Game/Database/RealmAccess.cs
+++ b/osu.Game/Database/RealmAccess.cs
@@ -13,6 +13,7 @@
 using System.Reflection;
 using System.Threading;
 using System.Threading.Tasks;
+using osu.Framework;
 using osu.Framework.Allocation;
 using osu.Framework.Development;
 using osu.Framework.Input.Bindings;
@@ -293,7 +294,17 @@ private void cleanupPendingDeletions()
         /// Compact this realm.
         /// </summary>
         /// <returns></returns>
-        public bool Compact() => Realm.Compact(getConfiguration());
+        public bool Compact()
+        {
+            try
+            {
+                return Realm.Compact(getConfiguration());
+            }
+            catch (AggregateException ae) when (RuntimeInfo.OS == RuntimeInfo.Platform.macOS && ae.Flatten().InnerException is TypeInitializationException)
+            {
+                return true;
+            }
+        }

         /// <summary>
         /// Run work on realm with a return value.
@@ -542,6 +553,10 @@ private Realm getRealmInstance()

                 return Realm.GetInstance(getConfiguration());
             }
+            catch (AggregateException ae) when (RuntimeInfo.OS == RuntimeInfo.Platform.macOS && ae.Flatten().InnerException is TypeInitializationException)
+            {
+                return Realm.GetInstance();
+            }
             finally
             {
                 if (tookSemaphoreLock)
```

Unsure if the patch is tolerable at all to handle such cases, but going to leave it here.

---

Other than the above, this is working as expected upon testing.

**Comment by peppy** [comment]
Probably fine, but they will need local `TODO: REMOVE ME` because the logic is now outside the migration process itself.

**Review by frenzibyte — APPROVED** [review]

**Pull request merged**